db_revisions/versions/6613e1e2c133_replace_is_active_with_lei_status_in_fi.py

```diff
@@ -0,0 +1,101 @@
+"""Replace is_active with lei_status in financial_institutions table
+
+Revision ID: 6613e1e2c133
+Revises: 6dd77f09fae6
+Create Date: 2024-11-13 00:43:53.489086
+
+"""
+
+from typing import Sequence, Union
+
+from alembic import op
+import sqlalchemy as sa
+
+from db_revisions.utils import table_exists, get_table_by_name, get_indices_from_collection
+
+
+# revision identifiers, used by Alembic.
+revision: str = "6613e1e2c133"
+down_revision: Union[str, None] = "6dd77f09fae6"
+branch_labels: Union[str, Sequence[str], None] = None
+depends_on: Union[str, Sequence[str], None] = None
+
+lei_status_seed_data = [
+    {"code": "ISSUED", "name": "Issued", "can_file": True},
+    {"code": "LAPSED", "name": "Lapsed", "can_file": False},
+    {"code": "RETIRED", "name": "Retired", "can_file": False},
+]
+
+
+def upgrade() -> None:
+
+    # Creating lei_status table
+    if not table_exists("lei_status"):
+        op.create_table(
+            "lei_status",
+            sa.Column("code", sa.String(), nullable=False),
+            sa.Column("name", sa.String(), nullable=False),
+            sa.Column("can_file", sa.Boolean(), nullable=False),
+            sa.Column("event_time", sa.DateTime(), server_default=sa.func.now(), nullable=False),
+            sa.PrimaryKeyConstraint("code"),
+            sa.UniqueConstraint("name"),
+        )
+
+    # Seeding lei_status table
+    lei_status_table = get_table_by_name("lei_status")
+    op.bulk_insert(lei_status_table, lei_status_seed_data)
+
+    # Removing is_active from and adding lei_status_code to financial_institutions table
+    op.add_column("financial_institutions", sa.Column("lei_status_code", sa.String(), nullable=True))
+
+    op.execute(
+        "UPDATE financial_institutions SET lei_status_code = CASE is_active WHEN True THEN 'ISSUED' ELSE 'LAPSED' END"
+    )
+
+    with op.batch_alter_table("financial_institutions") as batch_op:
+        batch_op.alter_column("lei_status_code", existing_nullable=True, nullable=False)
+        batch_op.create_index(
+            index_name=batch_op.f("ix_financial_institutions_lei_status_code"),
+            columns=["lei_status_code"],
+            unique=False,
+        )
+        batch_op.create_foreign_key(
+            "fk_lei_status_financial_institutions",
+            "lei_status",
+            ["lei_status_code"],
+            ["code"],
+        )
+
+        batch_op.drop_index(index_name="ix_financial_institutions_is_active")
+        batch_op.drop_column("is_active")
+
+    # Removing is_active from and adding lei_status_code to financial_institutions_history table
+    op.add_column("financial_institutions_history", sa.Column("lei_status_code", sa.String()))
+
+
+def downgrade() -> None:
+
+    op.drop_column("financial_institutions_history", "lei_status_code")
+
+    op.add_column(
+        "financial_institutions",
+        sa.Column(name="is_active", type_=sa.Boolean(), server_default=sa.true()),
```

Comment:
I wouldn't set a default for this; this would carry over into new institution creation if is_active field isn't specified, right? It should always be provided at creation time. For the downgrade purpose, I would have this as a nullable field, then update the existing records based on lei_status, then make this nullable = false.

Reply:
But is_active being set as required makes it so an institution can't be added without providing one.

Reply:
Nullable is just during this process; then it should be non-nullable at the completion of the process. It was an required field that needed to be provided. Although... since we're moving to a new paradigm, it really isn't relevant anymore. I guess this is inconsequential now; leave it as is is fine.